runtime/ops/pooling.cc

```diff
@@ -135,6 +135,132 @@ chainerx::Array ROIPool2D(const chainerx::Array& bottom_data, const chainerx::Ar
+
+    const int64_t channels = bottom_data.shape()[1];
+    const int64_t height = bottom_data.shape()[2];
+    const int64_t width = bottom_data.shape()[3];
+    const int64_t n_rois = bottom_rois.shape()[0];
+    const int64_t pooled_height = output_shape[0];
+    const int64_t pooled_width = output_shape[1];
+    chainerx::Array top_data = chainerx::Zeros(chainerx::Shape{n_rois, channels, pooled_height, pooled_width}, bottom_rois.dtype());
+
+    for(int64_t n = 0; n < n_rois; ++n) {
+        int64_t roi_batch_ind = int64_t(chainerx::AsScalar(bottom_roi_indices.At({n})));
+        double roi_start_h = double(chainerx::AsScalar(bottom_rois.At({n, 0})) * spatial_scale);
+        double roi_start_w = double(chainerx::AsScalar(bottom_rois.At({n, 1})) * spatial_scale);
+        double roi_end_h = double(chainerx::AsScalar(bottom_rois.At({n, 2})) * spatial_scale);
+        double roi_end_w = double(chainerx::AsScalar(bottom_rois.At({n, 3})) * spatial_scale);
+        double roi_height = std::max<double>(roi_end_h - roi_start_h, 1.);
+        double roi_width = std::max<double>(roi_end_w - roi_start_w, 1.);
+        double bin_size_h = roi_height / pooled_height;
+        double bin_size_w = roi_width / pooled_width;
+
+        int64_t roi_bin_grid_h = sampling_ratio[0];
+        int64_t roi_bin_grid_w = sampling_ratio[1];
+
+        for(int64_t c = 0; c < channels; ++c) {
+            for (int64_t ph = 0; ph < pooled_height; ++ph) {
+                for (int64_t pw = 0; pw < pooled_width; ++pw) {
+                    ReduceMode reduce;
+                    int64_t iy = 0;
+                    while(iy < roi_bin_grid_h) {
+                        auto y = roi_start_h + ph * bin_size_h + (iy + 0.5) * bin_size_h / roi_bin_grid_h;
+                        int64_t ix = 0;
+                        while(ix < roi_bin_grid_w) {
+                            auto x = roi_start_w + pw * bin_size_w + (ix + 0.5) * bin_size_w / roi_bin_grid_w;
+
+                            // bilinear interpolation {{
+                            auto params = get_bilinear_interp_params(y, x, height, width);
+                            if(!params) {
+                                continue;
+                            }
+                            int64_t y_low, x_low, y_high, x_high;
+                            double w1, w2, w3, w4;
+                            std::tie(y_low, x_low, y_high, x_high, w1, w2, w3, w4) = *params;
+                            auto v1 = float(chainerx::AsScalar(bottom_data.At({roi_batch_ind, c, y_low, x_low})));
+                            auto v2 = float(chainerx::AsScalar(bottom_data.At({roi_batch_ind, c, y_low, x_high})));
+                            auto v3 = float(chainerx::AsScalar(bottom_data.At({roi_batch_ind, c, y_high, x_low})));
+                            auto v4 = float(chainerx::AsScalar(bottom_data.At({roi_batch_ind, c, y_high, x_high})));
+                            reduce.Reduce(v1, v2, v3, v4, w1, w2, w3, w4);
```

Comment:
Perhaps it's better to calculate `w1 * v1 + w2 * v2 + w3 * v3 + w4 * v4` here and simplify the interface of `Reduce`?

Reply:
I see. I'll fix